### PR TITLE
[FLINK-25023] Prevent leaking classloaders from filesystems through AccessControlContext.

### PR DESCRIPTION
**Pull request by dmvk**
https://issues.apache.org/jira/browse/FLINK-25023

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java [hunk 1]
```diff
@@ -18,29 +18,52 @@
...
- * {@link FileSystem} operations.
+ * A wrapper around {@link FileSystemFactory} that ensures the same classloader is used for all
+ * {@link FileSystem} operations and that we don't leak classloader via inherited {@link
+ * java.security.ProtectionDomain protection domains} when new threads that might outlive the
+ * classloader are spawned during {@link FileSystem} initialization.
  */
-public class PluginFileSystemFactory implements FileSystemFactory {
+public class SafeFileSystemFactory implements FileSystemFactory {
+
+    private static final ExecutorService EXECUTOR =
```
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java:41]: We use the `ExecutorService` here (and instantiate a `Thread` right-away) so that the thread is created outside of the user-code context? This way, we ensure that the `FileSystem` is not loaded in user-code context. Does my reasoning match your intention?

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.core.fs;
+
+import org.apache.flink.util.ChildFirstClassLoader;
+import org.apache.flink.util.TemporaryClassLoaderContext;
+
+import org.junit.jupiter.api.Assertions;
```
Comment (XComp) [inline flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:23]: nit: based on the [newly updated guidelines](https://flink.apache.org/contributing/code-style-and-quality-common.html#testing), we should use `AssertJ` instead of JUnit asserts to help aligning the test infrastructure in this sense

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+import java.security.AccessControlContext;
+import java.security.AccessController;
+import java.security.ProtectionDomain;
+import java.util.List;
+import java.util.Vector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** Tests for the {@link SafeFileSystemFactory}. */
+public class SafeFileSystemFactoryTest {
```
Comment (XComp) [inline flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:42]: Maybe, it's the time of day, but I struggle to follow what the test is doing. I'm gonna try to go over it once more tomorrow morning. But it might be an indication that we could add a few more comments?

File: flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemTest.java [hunk 4]
```diff
@@ -97,10 +98,15 @@ public void testShadingOfAwsCredProviderConfig() {
     //  utilities
     // ------------------------------------------------------------------------
 
-    private static void validateBasicCredentials(FileSystem fs) throws Exception {
-        assertTrue(fs instanceof FlinkS3FileSystem);
+    private static void validateBasicCredentials(FileSystem wrappedFileSystem) throws Exception {
+        assertTrue(wrappedFileSystem instanceof WrappingProxy);
```
Comment (XComp) [inline flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemTest.java:102]: Why is this change needed now? Even in the previous version we used `ClassLoaderFixingFileSystem` which is where, I guess, the `WrappingProxy` is coming from. 🤔

**Review by XComp — COMMENTED** [review]
Thanks for providing this fix, @dmvk . Please see my questions below...

File: flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java [hunk 5]
```diff
@@ -326,14 +326,18 @@ public static void initialize(Configuration config, PluginManager pluginManager)
             FS_FACTORIES.clear();
 
             Collection<Supplier<Iterator<FileSystemFactory>>> factorySuppliers = new ArrayList<>(2);
-            factorySuppliers.add(() -> ServiceLoader.load(FileSystemFactory.class).iterator());
+            factorySuppliers.add(
```
[the remark this answers is not in the document]
Comment (dmvk) [reply flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java:329]: yes, but we're also extending the safeguard so it guarded against potential leak of protection domains

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java (hunk 2 above)
Comment (dmvk) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:23]: 👍

File: flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemTest.java (hunk 4 above)
Comment (dmvk) [reply flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemTest.java:102]: This FileSystem didn't come from plugin loader, so it hasn't been wrapped until this change

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java [hunk 6]
```diff
@@ -87,10 +91,8 @@ public static void executeProgram(
             boolean suppressSysout)
             throws ProgramInvocationException {
         checkNotNull(executorServiceLoader);
-        final ClassLoader userCodeClassLoader = program.getUserCodeClassLoader();
-        final ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
-        try {
-            Thread.currentThread().setContextClassLoader(userCodeClassLoader);
+        try (final TemporaryClassLoaderContext ignored =
+                TemporaryClassLoaderContext.of(program.getUserCodeClassLoader())) {
```
[the remark this answers is not in the document]
Comment (dmvk) [reply flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:95]: Yes, it's doing exactly the same thing. Also I had a desperate need to get rid of that ENUM 😇

File: flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java (hunk 1 above)
Comment (dmvk) [reply flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java:41]: This is actually the main point of this change.

First off, we need to understand a bit more about how the classloading works.

Every class that is loaded is assigned a protection domain. The protection domain contains reference to a CodeSource ("class location, ie. path to jar" + signing certificates), reference to a ClassLoader that has loaded it, principals and  permissions.

The part that bothers us the most is the reference to the ClassLoader that has loaded the class. There is no way to get rid of it, that wouldn't involve re-implementing a good part of the low level ClassLoaders (because everything is marked as final...).

---

Now to the actual problem. There is a thing called `AccessControlContext`, which can be obtained by calling `AccessController.getContext()`. This context contains list of current protection domains.

In pseudo code, the list of protection domain is obtained as  follows:

```
getCurrentStackTrace() | getStackTraceElementClass | Class#getProtectionDomain | uniq (as in bash)
```

When a new thread is spawned it inherits the `AccessControlContext` of the calller.

So... if we spawn any thread that outlives a job from a place that has been called by user code (there is anything loaded by user classloader on stack), we'll leak the reference to user classloader to this newly created thread.

This happens especially with hadoop filesystem, where the initialization of the filesystem spawn some "cleaner threads" that are attached to the JVM lifecycle.

In case of web-sumission, file system initialization may be actually called from the user code (see the stacktrace - https://gist.github.com/dmvk/a77207634cde8094baf456430d1bd92d).

---

The way to work around this, is to spawn initialize filesystems in a "safe" thread, that has no potentially leaky protection domains, that the new thread could inherit.

This is unfortunately the cleanest solution I was able to come up with after 3 days of research.

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java (hunk 3 above)
Comment (dmvk) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:42]: See https://github.com/apache/flink/pull/17916#discussion_r761288714 for  explanation.

Once we're sure that this is the approach we want to take, I'll try to come up with more explanatory javadocs.

We should also get more people to validate that we really want to do this before merging. We're still not sure with @zentol that the outcome outweights the complexity of this change.

File: flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemTest.java (hunk 4 above)
Comment (XComp) [reply flink-filesystems/flink-s3-fs-presto/src/test/java/org/apache/flink/fs/s3presto/PrestoS3FileSystemTest.java:102]: Ah, good point. I forgot that we added the safe guard also to the default filesystem loading when writing this comment. 👍

**Comment by dmvk** [comment]
@AHeise It would great to hear your thoughts on this one 🙏

File: flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java (hunk 1 above)
Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java:41]: What you're describing makes sense. I struggled to find documentation about where the `AccessControlContext` is coming from when creating a new `Thread`.

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java (hunk 3 above)
Comment (XComp) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:44]: What are your concerns about the approach you implemented in this PR? 🤔

Comment (dmvk) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:44]: Mostly it's about... lot of hustle for a "minor" issue. We'll leak at most one class loader per process without this fix, which should be bearable as it adds a constant overhead to the memory requirements.

But on the other hand, we shouldn't ideally leak anything and the final solution is not as "hacky" as the previous approaches. 🤔

Comment (XComp) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:44]: I wouldn't consider it "too hacky" if we add some more documentation. 👍

File: flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java [hunk 7]
```diff
@@ -48,11 +49,14 @@
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
 /** Utility functions for Flink client. */
-public enum ClientUtils {
-    ;
+public class ClientUtils {
```
Comment (AHeise) [inline flink-clients/src/main/java/org/apache/flink/client/ClientUtils.java:52]: @zentol ? ;)

File: flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java [hunk 8]
```diff
@@ -18,29 +18,52 @@
...
 /**
- * A wrapper around {@link FileSystemFactory} that ensures the plugin classloader is used for all
- * {@link FileSystem} operations.
+ * A wrapper around {@link FileSystemFactory} that ensures the same classloader is used for all
+ * {@link FileSystem} operations and that we don't leak classloader via inherited {@link
+ * java.security.ProtectionDomain protection domains} when new threads that might outlive the
+ * classloader are spawned during {@link FileSystem} initialization.
  */
-public class PluginFileSystemFactory implements FileSystemFactory {
+public class SafeFileSystemFactory implements FileSystemFactory {
```
Comment (AHeise) [inline flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java:39]: Note sure if the rename is necessary. I'd argue that your change also works for a `PluginFileSystemFactory`.

File: flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java [hunk 9]
```diff
@@ -58,8 +81,29 @@ public void configure(final Configuration config) {
 
     @Override
     public FileSystem create(final URI fsUri) throws IOException {
-        try (TemporaryClassLoaderContext ignored = TemporaryClassLoaderContext.of(loader)) {
-            return new ClassLoaderFixingFileSystem(inner.create(fsUri), loader);
+        final CompletableFuture<FileSystem> fileSystemFuture = new CompletableFuture<>();
+        EXECUTOR.execute(
+                () -> {
+                    try {
+                        synchronized (inner) {
```
Comment (AHeise) [inline flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java:88]: Why do we need to sync here? The same thread can only have one context loader...

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                    Class.forName(ProtectionDomainLeakTest.class.getName(), true, leakyClassLoader);
+            final Constructor<?> declaredConstructor =
+                    clazz.getDeclaredConstructor(FileSystemFactory.class);
+            declaredConstructor.setAccessible(true);
+            declaredConstructor.newInstance(factory);
+        }
+        Assertions.assertEquals(1, getLoadedFlinkClasses(leakyClassLoader).size());
+        Assertions.assertEquals(
+                ProtectionDomainLeakTest.class.getName(),
+                getLoadedFlinkClasses(leakyClassLoader).get(0).getName());
```
Comment (AHeise) [inline flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java]: So the classloader is leaky in any ways and we really just try to contain the leak to CL and not also PD?

**Review by AHeise — COMMENTED** [review]
Interesting approach. Could you please split the PR to make it easier to get to the beef? 
I need a second pass, earlier in the day to fully understand it and the implications.

File: flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java (hunk 1 above)
Comment (dmvk) [reply flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java:41]: I'll add comments into the code.

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java (hunk 10 above)
Comment (dmvk) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java]: yes, it's all about leaking the reference to the classloader, which makes it ineligible for the GC once the job finishes.

**Comment by dmvk** [comment]
@XComp ready for another pass

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                            // No-op.
+                        });
+        final Class<?> clazz =
+                Class.forName(ProtectionDomainLeakTest.class.getName(), true, leakyClassLoader);
+        final Constructor<?> declaredConstructor =
+                clazz.getDeclaredConstructor(FileSystemFactory.class);
+        declaredConstructor.setAccessible(true);
+        declaredConstructor.newInstance(factory);
+        // We have to assert class name here, because the class we want to assert has been loaded in
+        // a different classloader.
```
Comment (XComp) [inline flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java:146]: Initially, I didn't understand this comment here. Not sure whether it would help others as well mentioning the implicit fact that the `Class` object attached to an instance of the class is actually a new `Class` instance when being loaded from a different classloader

File: flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java [hunk 12]
```diff
@@ -18,29 +18,85 @@
...
 import java.net.URI;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
 
 /**
- * A wrapper around {@link FileSystemFactory} that ensures the plugin classloader is used for all
- * {@link FileSystem} operations.
+ * A wrapper around {@link FileSystemFactory} that ensures the same classloader is used for all
```
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/core/fs/SafeFileSystemFactory.java:34]: There seems to be some checkstyle issue with this JavaDoc.

**Review by XComp — CHANGES_REQUESTED** [review]
The build didn't go through due to checkstyle concerns. This needs to be fixed. Otherwise, I have only minor comments.

I checked the tests verifying that the `SafeFileSystemFactoryTest` is failing without the change. Additionally, I went through the test code once more trying to reason it and couldn't find anything suspicious.

File: flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java (hunk 10 above)
Comment (XComp) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java]: Out of curiosity: CL means class loading, doesn't it? What about PD? 🤔

Comment (zentol) [reply flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java]: PD=ProtectionDomain
